lib/server.js

```diff
@@ -0,0 +1,131 @@
+var assert = require('assert');
+
+var dockerSocket = '/var/run/docker.sock';
+var stats = fs.statSync(dockerSocket);
+if (!stats.isSocket()) {
+  throw new Error('Are you sure the docker is running?');
+}
+var docker = new Docker({socketPath: dockerSocket});
+
+var dbgstream;
+if (process.argv[2] === '--dump') {
+  dbgstream = fs.createWriteStream(path.join(__dirname, '/dump.log'));
+}
+
+// Makes a new buffer with code number (ideally 0x00-0xFF)
+// appended to the start of the given buffer
+function muxBuffer (code, buffer) {
+  return Buffer.concat([new Buffer(String.fromCharCode(code)), buffer]);
+}
+
+/** Makes an instance of docker exec running command, then links the provided
+    websocket to the stream provided by docker exec
+
+    @param tty: boolean representing whether or not terminal emulation is active
+  */
+function execWithSocket (container, socket, command, tty) {
+  var options = {
+    AttachStdout: true,
+    AttachStderr: true,
+    AttachStdin: true,
+    Tty: tty,
+    Detach: false,
+    Cmd: command,
+  };
+  var attachoptions = {
+    stdin: true,
+    stdout: true,
+    stderr: true,
+    stream: true,
+  };
+  return container.exec(options, (err, exec) => {
+    if (err) { //err proof
+      throw err;
+    }
+    exec.start(attachoptions, (err, execStream) => {
+      if (err) { //err proof
+        throw err;
+      }
+      var strout = through((data) => {
+        if (dbgstream) {
+          dbgstream.write('OUT: ' + data + '\n-\n');
+        }
+        socket.send(muxBuffer(1, data), {binary: true, mask: true});
+      });
+      var strerr = through((data) => {
+        if (dbgstream) {
+          dbgstream.write('ERR: ' + data + '\n-\n');
+        }
+        socket.send(muxBuffer(2, data), {binary: true, mask: true});
+      });
+      exec.modem.demuxStream(execStream, strout, strerr);
+
+      socket.on('message', (message) => {
+        if (dbgstream) {
+          dbgstream.write('IN: ' + message + '\n-\n');
+        }
+        execStream.write(message);
+      });
+
+      socket.on('disconnect', () => {
+        execStream.end('\x03\x04\r\nexit\r\n'); //should be how to ctrl+c ctrl+d, might be better way to kill
+      });
+
+      execStream.on('end', () => {
+        exec.inspect((err, data) => {
+          if (err) {
+            throw err;
+          }
+          var exitBuf = new Buffer(4);
+          exitBuf.writeUInt32LE(data.ExitCode);
+          socket.send(muxBuffer(100, exitBuf));
+        });
+      });
+
+      socket.send('ready'); //send ready message to let client know it is ready
+    });
+  });
+}
+
+export default class DockerExecWebsocketServer {
+  /* Creates Docker Exec instance on given container, running the first message given
+   * as a command.
+   * Options:
+   * port, required
+   * container, name or id of docker container, required
+   * path, path where the websocket is hosted
+   * tty, whether or not we expect VT100-style output
+   */
+  constructor (options) {
```

Comment:
In general for this function... Please add some empty link breaks... splitting things into sections of code.
Like:
```js
    options = _.defaults({}, options, {
      path: '/'+slugid.v4(),
      tty: false
    });
    assert(this.port, 'required port option missing');
    assert(container, 'required container option missing');

    // Get container wrapper
    var container = docker.getContainer(options.container);

    // Store port and path
    this.port = options.port;
    this.path = options.path;

    // Create server
    this.server = new ws.Server({
      port: this.port,
      path: this.path,
    });
    debug('%s%s created', this.port, this.path);
```

Also when you use `debug`, please don't do `+` instead of doing `debug(wsopts.port+wsopts.path+' created');` do `debug('%s%s created', wsopts.port, wsopts.path);` they way the string concatenation happens inside the `debug` function, and if the `DEBUG` environment variables isn't set, the `debug` function will ignore its arguments, and no overhead will be paid.